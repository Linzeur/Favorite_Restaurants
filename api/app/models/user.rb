class User < ApplicationRecord

  validates :name, presence: true
  validates :email, format: { with: /[a-zA-Z0-9]*@[a-zA-Z]*.[a-zA-Z0-9]*/, message: "Bad format"  },
                    uniqueness: true

  has_secure_password
  has_secure_token
  has_secure_token :reset_digest

  has_many :locations
  has_many :favorite_restaurants

  def self.valid_login?(email, password)
    user = find_by(email: email)
    user if user&.authenticate(password)
  end

  def invalidate_token
    update(token: nil)
  end
end
