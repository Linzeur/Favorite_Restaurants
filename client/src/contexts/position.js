import React from "react";

const PositionContext = React.createContext();

function PositionProvider(props) {
  const [position, setPosition] = React.useState([0, 0, false]);

  React.useEffect(() => {
    const watchId = navigator.geolocation.watchPosition(({ coords }) => {
      setPosition([coords.latitude, coords.longitude, true]);
      sessionStorage.setItem(
        "lastCurrentPosition",
        JSON.stringify({ lat: coords.latitude, lng: coords.longitude })
      );
    });
    return () => navigator.geolocation.clearWatch(watchId);
  }, [setPosition]);

  const value = {
    latitude: position[0],
    longitude: position[1],
    callAPI: position[2],
    onChange: (lat, lng, callAPI) => {
      setPosition([lat, lng, callAPI]);
    }
  };

  return <PositionContext.Provider value={value} {...props} />;
}

export { PositionProvider, PositionContext };
