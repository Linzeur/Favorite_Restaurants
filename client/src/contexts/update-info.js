import React from "react";

const UpdateListContext = React.createContext();

function UpdateListProvider(props) {
  const [updateList, setUpdateList] = React.useState({
    locations: 0,
    favorites: 0
  });

  const value = {
    locations: updateList.locations,
    favorites: updateList.favorites,
    onChange: key => {
      setUpdateList(prevState => {
        return { ...prevState, [key]: prevState[key] + 1 };
      });
    }
  };

  return <UpdateListContext.Provider value={value} {...props} />;
}

export { UpdateListProvider, UpdateListContext };
