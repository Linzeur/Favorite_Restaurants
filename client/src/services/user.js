import { apiUrl } from "../util";

const API_USERS = `${apiUrl}users`;

async function createUser(userData) {
  const response = await fetch(API_USERS, {
    method: "POST",
    credentials: "include",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify(userData)
  });

  if (!response.ok) {
    const { errors } = await response.json();
    throw new Error(errors.message);
  }

  return response.json();
}

export { createUser };
