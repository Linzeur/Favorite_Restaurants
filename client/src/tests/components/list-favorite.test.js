import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";

import ListFavorites from "../../components/list-favorite";
import { UserProvider } from "../../contexts/user";
import { PositionProvider } from "../../contexts/position";
import {
  UpdateListProvider,
  UpdateListContext
} from "../../contexts/update-info";

const user = {
  name: "User Test",
  email: "test@test.com"
};

const listFavorites = [
  {
    id: 1,
    restaurant_id: "ChIJpSBYQ2TOBZERrK59vkz3v9g",
    name: "La Familia",
    latitude: "-11.9913905",
    longitude: "-77.0873197"
  },
  {
    id: 2,
    restaurant_id: "ChIJyTFGS4XPBZER3KOQZ5MtZEk",
    name: "PIZZERIA FAMILY VEGA",
    latitude: "-11.986539",
    longitude: "-77.0886132"
  },
  {
    id: 4,
    restaurant_id: "ChIJC5ldYmTIBZERTQdPOFCzsSY",
    name: "Alphonse Bar",
    latitude: "-12.0917299",
    longitude: "-77.0248756"
  }
];

test("list-favorite component cosuming API", async () => {
  localStorage.setItem("user", JSON.stringify(user));
  fetch.mockResponseOnce(JSON.stringify(listFavorites));
  const setOptionSelectedFn = jest.fn();

  const { getAllByLabelText, asFragment } = render(
    <UserProvider>
      <UpdateListProvider>
        <UpdateListContext.Consumer>
          {updateInfo => (
            <ListFavorites optionSelectedFn={setOptionSelectedFn} />
          )}
        </UpdateListContext.Consumer>
      </UpdateListProvider>
    </UserProvider>
  );

  let favorites;
  await wait(() => {
    favorites = getAllByLabelText("Element Favorite");
  });

  expect(asFragment()).toMatchSnapshot();

  expect(favorites.length).toBe(3);
});

test("select an option", async () => {
  localStorage.setItem("user", JSON.stringify(user));
  fetch.mockResponseOnce(JSON.stringify(listFavorites));
  const setOptionSelectedFn = jest.fn();

  const { getAllByLabelText, asFragment } = render(
    <UserProvider>
      <PositionProvider>
        <UpdateListProvider>
          <UpdateListContext.Consumer>
            {updateInfo => (
              <ListFavorites optionSelectedFn={setOptionSelectedFn} />
            )}
          </UpdateListContext.Consumer>
        </UpdateListProvider>
      </PositionProvider>
    </UserProvider>
  );

  let favorites;
  await wait(() => {
    favorites = getAllByLabelText("Element Favorite");
  });

  expect(asFragment()).toMatchSnapshot();

  expect(favorites.length).toBe(3);

  fireEvent.click(favorites[0]);

  expect(setOptionSelectedFn).toHaveBeenCalledWith(listFavorites[0].id);
});
