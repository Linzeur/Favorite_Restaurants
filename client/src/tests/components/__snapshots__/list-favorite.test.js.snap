// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list-favorite component cosuming API 1`] = `
<DocumentFragment>
  <div
    class="css-1ylqutl-Search"
  >
    <input
      aria-label="Search an option"
      class="css-x0timv-Search"
      id="search"
      name="search"
      placeholder="Search an option"
      type="text"
    />
  </div>
  <ul
    class="css-1h1c26y-ListFavorite"
  >
    <li
      class="css-70qvj9"
    >
      <div
        aria-label="Element Favorite"
        class="css-j2js4d"
        role="button"
      >
        <svg
          class="css-ywh1g1"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
          />
        </svg>
        <div
          class="css-esxnue"
        >
          <span>
            La Familia
          </span>
        </div>
      </div>
      <button
        aria-label="Delete favorite restaurant"
        class="css-10jcxoq"
        type="button"
      >
        <svg
          class="css-3fuzg5"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
      </button>
    </li>
    <li
      class="css-70qvj9"
    >
      <div
        aria-label="Element Favorite"
        class="css-j2js4d"
        role="button"
      >
        <svg
          class="css-ywh1g1"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
          />
        </svg>
        <div
          class="css-esxnue"
        >
          <span>
            PIZZERIA FAMILY VEGA
          </span>
        </div>
      </div>
      <button
        aria-label="Delete favorite restaurant"
        class="css-10jcxoq"
        type="button"
      >
        <svg
          class="css-3fuzg5"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
      </button>
    </li>
    <li
      class="css-70qvj9"
    >
      <div
        aria-label="Element Favorite"
        class="css-j2js4d"
        role="button"
      >
        <svg
          class="css-ywh1g1"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
          />
        </svg>
        <div
          class="css-esxnue"
        >
          <span>
            Alphonse Bar
          </span>
        </div>
      </div>
      <button
        aria-label="Delete favorite restaurant"
        class="css-10jcxoq"
        type="button"
      >
        <svg
          class="css-3fuzg5"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
      </button>
    </li>
  </ul>
  <div
    class="css-gn3t32-Footer"
  >
    <span
      class="css-8ep07x-Footer"
    >
      Results:
    </span>
    <span
      class="css-xp9py7-Footer"
    >
      3 favorites
    </span>
  </div>
</DocumentFragment>
`;

exports[`select an option 1`] = `
<DocumentFragment>
  <div
    class="css-1ylqutl-Search"
  >
    <input
      aria-label="Search an option"
      class="css-x0timv-Search"
      id="search"
      name="search"
      placeholder="Search an option"
      type="text"
    />
  </div>
  <ul
    class="css-1h1c26y-ListFavorite"
  >
    <li
      class="css-70qvj9"
    >
      <div
        aria-label="Element Favorite"
        class="css-j2js4d"
        role="button"
      >
        <svg
          class="css-ywh1g1"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
          />
        </svg>
        <div
          class="css-esxnue"
        >
          <span>
            La Familia
          </span>
        </div>
      </div>
      <button
        aria-label="Delete favorite restaurant"
        class="css-10jcxoq"
        type="button"
      >
        <svg
          class="css-3fuzg5"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
      </button>
    </li>
    <li
      class="css-70qvj9"
    >
      <div
        aria-label="Element Favorite"
        class="css-j2js4d"
        role="button"
      >
        <svg
          class="css-ywh1g1"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
          />
        </svg>
        <div
          class="css-esxnue"
        >
          <span>
            PIZZERIA FAMILY VEGA
          </span>
        </div>
      </div>
      <button
        aria-label="Delete favorite restaurant"
        class="css-10jcxoq"
        type="button"
      >
        <svg
          class="css-3fuzg5"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
      </button>
    </li>
    <li
      class="css-70qvj9"
    >
      <div
        aria-label="Element Favorite"
        class="css-j2js4d"
        role="button"
      >
        <svg
          class="css-ywh1g1"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
          />
        </svg>
        <div
          class="css-esxnue"
        >
          <span>
            Alphonse Bar
          </span>
        </div>
      </div>
      <button
        aria-label="Delete favorite restaurant"
        class="css-10jcxoq"
        type="button"
      >
        <svg
          class="css-3fuzg5"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
      </button>
    </li>
  </ul>
  <div
    class="css-gn3t32-Footer"
  >
    <span
      class="css-8ep07x-Footer"
    >
      Results:
    </span>
    <span
      class="css-xp9py7-Footer"
    >
      3 favorites
    </span>
  </div>
</DocumentFragment>
`;
